### ISSUE TITLE
JIT: improve aliasing model for implicit-by-ref struct passing

### ISSUE THREAD

**Issue by JosephTremoulet** [dotnet/coreclr#10049]
There are cases on both ARM64 and Windows x64 where the calling convention dictates that a struct which is passed by value at source be passed by reference (after making a copy in the caller's frame) in the generated code.  Today we do this rewrite in the importer, and the callee's view of such parameters is the same as normal byrefs.  This inhibits promoting these structs, removing redundant loads of their fields, etc.  We'd get better codegen if we either modeled these as by-value until after later in the phase list or introduced internal annotations in the jit so that we could model these byref pointers as both non-null and restrict.

**Comment by russellhadley** [comment]
Keep in 2.0 for Span on Kestrel.

**Comment by JosephTremoulet** [comment]
Seeing as how we don't do much optimization on structs that we don't promote, I think the task here should really be to enable promotion of implicit-by-ref struct parameters (which conceptually we could do by copying the parameter to a local struct on method entry, rewriting all uses of the parameter to be uses of the local, and promoting the local).

**Comment by JosephTremoulet** [comment]
Turns out we make struct promotion decisions before we do address-taken analysis; a promoted struct that's address-taken just has its fields laid out together on the frame and the codegen basically looks like it wasn't promoted.  So inserting a copy at function entry would just add a useless mem-to-mem copy in those cases.

**Comment by JosephTremoulet** [comment]
Looks like the address-taken analysis, implicit-by-ref marking, and struct promotion should be reorderable, so I'll try swapping them around and find out if I'm right about that.

**Comment by ahsonkhan** [comment]
@JosephTremoulet, any updates on this issue?

**Comment by JosephTremoulet** [comment]
Per my previous notes, I spent some time working on running things in this order:

  1. address-taken analysis
  2. implicit by-ref rewrite
  3. struct promotion

It ran into issues where we lost some optimization capability because we do field-sensitive address-taken analysis, which depends on running struct promotion first for the address-taken analysis' bookkeeping.  So I've switched and am working on running things in this order:

  1. struct promotion
  2. address-taken analysis
  3. implicit by-ref rewrite

I have that doing the right thing in some basic testcases, but am still working through broader correctness (currently hitting some issues with Vector types) and assessing CQ/TP impact.  (That work-in-progress is in #10453).
